### ISSUE TITLE
Create CscMatrix from a continuous slice

### ISSUE THREAD

**Issue by bytesnake** [oxfordcontrol/osqp.rs#8]
I'm trying to copy a symmetric matrix in `nalgebra` to a `CscMatrix`, both are in column major format. My current approach is straightforward copying:
```
(0..mat.nrows()).map(|i| mat.row(i).iter().cloned().collect()).collect()
```
But this should be rather slow, because the matrix is first transferred into row major format and then back again. Is there any way to construct a `CscMatrix` from a continuous slice?

**Comment by ebarnard** [comment]
It is possible to manually create a dense `CscMatrix` with something like the following:
```rust
CscMatrix {
    nrows: mat.nrows(),
    ncols: mat.ncols(),
    indptr: (0..mat.ncols()).map(|i| i * mat.nrows()).collect().into(),
    indices: (0..mat.ncols * mat.nrows().collect().into(),
    data: mat.as_slice().into(),
}
```
But note that you still have to allocate a `Vec` for the row and column indices as CSC matrices are designed to hold sparse matrices.

However, it shouldn't be the case that you have to regularly construct a `CscMatrix` containing owned data. Could you let me know a bit more about the overall problem you're trying to solve?

**Comment by bytesnake** [comment]
Hey, I'm trying to solve the dual problem of a basic SVM classifier with l1-regularization. The problem is stated in my class as
![image](https://user-images.githubusercontent.com/989353/56078942-dee9b000-5ded-11e9-9e54-1fc2271cc53b.png)
with lambda_i the dual variables, y_i classes and x_i feature vectors.
The matrix P in *osqp* is obviously a covariance matrix with some sign changes. It is also symmetric, but not a sparse matrix like `CscMatrix`. Should I tackle the problem in a different way? I have no idea how I should avoid constructing the `CscMatrix` explicitly, but I don't have much experience in convex optimization either.

**Comment by ebarnard** [comment]
OSQP is designed primarily with sparse problems in mind (e.g. MPC where large portions of P are always zero) which is why it requires `P` in CSC matrix form. But it does work perfectly well with dense problems.

The `P` matrix for SVM doesn't have any entries that are always zero so there's no general sparsity structure, but nonetheless you need to provide `P` in sparse CSC form. Moreover `P` depends on the SVM training data so it must be recalculated for each training set.

Unless your problem is very small the time taken for re-factorising `P` and performing the optimisation will far outweigh the time taken to perform the unnecessary allocations. Nonetheless I think they can be avoided by not collecting the intermediate results into vectors with something like:
```rust
CscMatrix::from((0..mat.nrows()).map(|i| mat.row(i).iter()))
```
which uses this impl:
```rust
impl<'a, I: 'a, J: 'a> From<I> for CscMatrix<'static>
where
    I: IntoIterator<Item = J>,
    J: IntoIterator<Item = &'a f64>,
```

If you're performing cross-validation with same-sized datasets you could reduce allocations further by keeping the `indptr` and `indices` vectors around between iterations and creating a `CscMatrix` for each training dataset that only contains `Cow::Borrowed` values.

**Comment by bytesnake** [comment]
Well, my primary goal is not speed, but clarity because I want to use this example in a tutorial. For my problem `P` is not sparse, unless I'm using a kernel function which can exploit a lot of sparsity in the feature data structure. 
Your example doesn't work, because of the borrowing system. The compiler can't interfere the lifetime of `row(i)`, but I couldn't come up with something better .. would you accept a PR, which just adds a constructor method for continuous dense data?

**Comment by ebarnard** [comment]
Ah well.

Yes, that would be great. I suppose something along the lines of nalgebra's `MatrixMN::from_iterator_generic`?

**Comment by bytesnake** [comment]
I created a PR with the new function ~ d7add0f8395c70aab4e47aa6eb1dd5a2b9962554

**Comment by ebarnard** [comment]
Changes merged in 64ad54784e10c36a6180799cc9735789e285e14c. Feel free to re-open if you still have issues.

cc #9
